### PR TITLE
Update Python to be minimum 3.9

### PR DESCRIPTION
**Pull request by ericmjl**
This PR updates our minimum Python version to be 3.9.

Key changes:

- We matrix Python version testing across 3.9, and 3.10 (current stable version and two before).
- We note everywhere relevant that the minimum Python version should be 3.9.

### CODE REVIEW

**Comment by ericmjl** [comment]
@Zeroto521 I just realized this PR might overlap with your other PR #1143. I'll just use the PR to test some syntax things needed to make Python >=3.8 work. After that, we can discard my PR, and you can pick-and-choose whatever changes seem to make sense into your PR.

**Comment by ericmjl** [comment]
🚀 Deployed on https://deploy-preview-1157--pyjanitor.netlify.app

**Comment by Zeroto521** [comment]
Directly requires py3.9. Would that be a little bit aggressive?
pandas main branch still requires py3.8.
I'm okay to deprecate py3.6 and py3.7.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -34,6 +35,7 @@ jobs:
       - name: Run unit tests
         run: |
           conda activate pyjanitor-dev
+          mamba install python=${{ matrix.python-version }}
```
Comment (Zeroto521) [inline .github/workflows/tests.yml:38]: those all cis actually use python 3.9.
https://github.com/pyjanitor-devs/pyjanitor/runs/7942582404?check_suite_focus=true#step:3:905

**Comment by ericmjl** [comment]
Good question, @Zeroto521. Type annotations in Python 3.9 are more flexible, i.e. we can use built-in `tuple` or `list` as types like this:

```python
def func(df, arg1: tuple[int]):
    pass 
```

We can drop down to Python 3.8 if we fix the types to be:

```python
from typing import Tuple # (and/or List)

def func(df, arg1: Tuple[int]):
    pass
```

**Comment by Zeroto521** [comment]
Py3.7+ with `__future__` also has that feature.

```python
from __future__ import annotations

int_array: list[int] = list(range(3))
None_or_bool: bool | None = True  # equals to `Optional[bool]`
```

**Comment by thatlittleboy** [comment]
I also suggest using asottile's [`pyupgrade`](https://github.com/asottile/pyupgrade) to upgrade whatever syntax we can to the minimum version we are supporting. I also suggest sticking to py38 for now instead of jumping straight to py39. py37 EOL is coming this year.

Less manual work :)

**Comment by ericmjl** [comment]
@thatlittleboy I'm in favour of more automation! 😺

Lemme take care of that in a bit.

**Comment by ericmjl** [comment]
This PR is going stale. I'm going to close it and re-do it with minimum py38, as per @Zeroto521's and @thatlittleboy's suggestions.

**Pull request closed without merging**